File: vibeshield-ts/src/middleware/ssrf-security.ts
import { validateUrl } from '../core/ssrf-protector.js';

/**
 * Custom error thrown when an SSRF validation check fails.
 */
export class VibeShieldSSRFError extends Error {
  public url: string;
  public reason: string;

  constructor(url: string, reason: string) {
    super(`[VibeShield] SSRF Protection: Blocked URL "${url}". Reason: ${reason}`);
    this.name = 'VibeShieldSSRFError';
    this.url = url;
    this.reason = reason;
    Object.setPrototypeOf(this, VibeShieldSSRFError.prototype);
  }
}

/**
 * Enforces that the given URL is safe from SSRF attacks.
 * Throws a VibeShieldSSRFError if the URL is deemed unsafe.
 *
 * @param targetUrl - The target URL to validate.
 * @throws {VibeShieldSSRFError} If the URL is not safe.
 */
export function enforceSafeUrl(targetUrl: string): void {
  const result = validateUrl(targetUrl);
  if (!result.safe) {
    throw new VibeShieldSSRFError(targetUrl, result.reason || 'Blocked by SSRF policy');
  }
}

File: vibeshield-python/src/middleware/ssrf_security.py
"""
VibeShield SSRF Security Middleware

Enforces SSRF URL validation safety checks and raises custom exceptions.
"""

import logging
from ..ssrf_protector import validate_url

logger = logging.getLogger("VibeShield")


class VibeShieldSSRFError(ValueError):
    """
    Custom exception raised when a URL is blocked by VibeShield's SSRF policy.
    """
    def __init__(self, url: str, reason: str):
        super().__init__(f"[VibeShield] SSRF Protection: Blocked URL \"{url}\". Reason: {reason}")
        self.url = url
        self.reason = reason


def enforce_safe_url(target_url: str) -> None:
    """
    Validates a URL and raises an exception if it is not safe.

    Args:
        target_url: The URL string to validate.

    Raises:
        VibeShieldSSRFError: If the URL fails SSRF safety validation.
    """
    result = validate_url(target_url)
    if not result['safe']:
        reason = result.get('reason', 'Blocked by SSRF policy')
        logger.warning("[VibeShield] SSRF safety violation: Blocked URL \"%s\". Reason: %s", target_url, reason)
        raise VibeShieldSSRFError(target_url, reason)
